docs/messaging/messaging-spans.md

```diff
@@ -196,21 +198,77 @@ Examples:
+| `deliver`      | One or more messages are passed to a consumer. This operation refers to push-based scenarios, where consumer register callbacks which get called by messaging SDKs. |
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+For cases not covered by the table above, the span kind should be set according
+to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to CLIENT for the "Publish" span if its context is not
+used as creation context and if the "Publish" span models a synchronous call to
+the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always exist between a PRODUCER span on the producer
+side and a CONSUMER span on the consumer side. This allows analysis tools to
+interpret linked traces without the need for additional semantic hints.
+
+### Trace structure
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a user provides a custom creation context in a message, this context SHOULD
+NOT be modified, a "Create" span SHOULD NOT be created, and the "Publish" span
+SHOULD link to the custom creation context.  Otherwise, if a "Create" span
+exists for a message, its context SHOULD be injected into the message. If no
+"Create" span exists and no custom creation context is injected into the
+message, the context of the related "Publish" span SHOULD be injected into the
+message.
+
+#### Consumer spans
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application code
+(push-based scenarios).
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code
+(pull-based scenarios).
+
+"Deliver" or "Receive" spans MUST NOT be created for messages that are
+pre-fetched or cached by messaging libraries or SDKs until they are forwarded
+to the caller.
+
+A single "Deliver" or "Receive" span can account for a single message, for a
```

Comment:
I think we're missing something that talks about deliver span duration from the OTEP: "A "Deliver" span covers the call of such a callback or handler"

Reply:
I added this to the first paragraph:

> "Deliver" spans SHOULD be created for operations of passing messages to the application when those operations are not initiated by the application code (push-based scenarios). _A "Deliver" span covers the duration of such an operation, which is usually a callback or handler._